clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2918,24 +2919,60 @@ static bool checkWorkGroupSizeValues(Sema &S, Decl *D, const ParsedAttr &Attr,
   };
 
   if (Attr.getKind() == ParsedAttr::AT_SYCLIntelMaxGlobalWorkDim) {
-    if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>())
-      Result &= checkZeroDim(A, A->getXDim(), A->getYDim(), A->getZDim());
-    if (const auto *A = D->getAttr<ReqdWorkGroupSizeAttr>())
-      Result &= checkZeroDim(A, A->getXDim(), A->getYDim(), A->getZDim());
+    if (const auto *A = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+      Optional<llvm::APSInt> XDimVal =
+          A->getXDim()->getIntegerConstantExpr(S.getASTContext());
+      Optional<llvm::APSInt> YDimVal =
+          A->getYDim()->getIntegerConstantExpr(S.getASTContext());
+      Optional<llvm::APSInt> ZDimVal =
+          A->getZDim()->getIntegerConstantExpr(S.getASTContext());
+      Result &= checkZeroDim(A, XDimVal->getZExtValue(),
+                             YDimVal->getZExtValue(), ZDimVal->getZExtValue());
+    } else if (const auto *A = D->getAttr<ReqdWorkGroupSizeAttr>()) {
+      Optional<llvm::APSInt> XDimVal =
+          A->getXDim()->getIntegerConstantExpr(S.getASTContext());
+      Optional<llvm::APSInt> YDimVal =
+          A->getYDim()->getIntegerConstantExpr(S.getASTContext());
+      Optional<llvm::APSInt> ZDimVal =
+          A->getZDim()->getIntegerConstantExpr(S.getASTContext());
+      Result &= checkZeroDim(A, XDimVal->getZExtValue(),
+                             YDimVal->getZExtValue(), ZDimVal->getZExtValue());
+    }
     return Result;
   }
 
   if (const auto *A = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
     int64_t AttrValue =
         A->getValue()->getIntegerConstantExpr(S.Context)->getSExtValue();
-    if (AttrValue == 0)
-      Result &= checkZeroDim(A, WGSize[0], WGSize[1], WGSize[2],
+    if (AttrValue == 0) {
+      Optional<llvm::APSInt> AttrXDim =
+          Attr.getArgAsExpr(2)->getIntegerConstantExpr(S.getASTContext());
```

Comment:
Are these variables named correctly? This is getting the third argument to the attribute and calling it "XDim", which is the first argument to the attribute.

Assuming the code is incorrect, adding some test coverage for this case may not be a bad idea.

Reply:
Inside handleWorkGroupSize(), we have check like this.
if (S.getLangOpts().SYCLIsDevice)
    std::swap(XDimExpr, ZDimExpr);

The Value of XDim() and ZDim() gets swapped, so for checkWorkGroupSizeValues() function, we will have XDim that will take third argument to the attribute  and ZDim will take first argument. @AaronBallman,does it make sense to you?

Reply:
In some regards, yes it makes sense, but in other regards, not as much. ;-) Why are we swapping the dimensions in the first place?

I guess my intuition would be that the semantic attribute would have kept the arguments in the same order as they were parsed in, and when performing code generation, we would change the argument order when lowering, as needed.

I hadn't yet commented on the swap because I was still trying to think through it, but I think one problem from that approach is that a user who writes the order as X, Y, Z in their source will get Z, Y, X when they pretty print the AST (for instance), unless we undo that swap. The same will happen when dumping the AST to text or JSON.

Reply:
>>Why are we swapping the dimensions in the first place?
Not sure, i did not implement the original implementation. 

yes, i agree, it is very confusing (swapping the value). AST prints different orders (Z, Y, X) than the user writes the order as X, Y, Z

Reply:
@AaronBallman, i have found this information about why we are swapping dimension here:
To follow the SYCL/C++ convention(SYCL 1.2.1), where the rightmost index maps to the linear dimension, values are swapped here.  OpenCL convention, where the leftmost index maps to the linear dimension. PR: https://github.com/intel/llvm/pull/1234

Reply:
Thanks for the further information!

If the attribute is only used in doing semantic checks for applying other attributes and when lowering to LLVM IR, then I think the swap should happen when lowering to LLVM IR and not when creating the semantic attribute.

If the attribute is used in other parts of the compiler (static analysis checks, clang-tidy checks, other parts of Sema than SemaDeclAttr.cpp), then I think it's worth discussing whether the semantic attribute should have the swapped values (so that users of the semantic attribute don't have to think about whether they need the swap or not). In this case, we should probably pick more descriptive names than X, Y, and Z for the semantic attribute arguments (and we'd have to add some code in to the various AST places to unswap).

WDYT?

Reply:
@MrSidims could you please comment on this?

Reply:
> Thanks for the further information!
> 
> If the attribute is only used in doing semantic checks for applying other attributes and when lowering to LLVM IR, then I think the swap should happen when lowering to LLVM IR and not when creating the semantic attribute.
> 
> If the attribute is used in other parts of the compiler (static analysis checks, clang-tidy checks, other parts of Sema than SemaDeclAttr.cpp), then I think it's worth discussing whether the semantic attribute should have the swapped values (so that users of the semantic attribute don't have to think about whether they need the swap or not). In this case, we should probably pick more descriptive names than X, Y, and Z for the semantic attribute arguments (and we'd have to add some code in to the various AST places to unswap).
> 
> WDYT?

The attribute is also applied to SYCL kernel, This is used in Sema part (SemaDeclAttr.cpp, SemaDecl.cpp, and SemaSYCL.cpp file)

Reply:
> Thanks for the further information!
> 
> If the attribute is only used in doing semantic checks for applying other attributes and when lowering to LLVM IR, then I think the swap should happen when lowering to LLVM IR and not when creating the semantic attribute.
> 
> If the attribute is used in other parts of the compiler (static analysis checks, clang-tidy checks, other parts of Sema than SemaDeclAttr.cpp), then I think it's worth discussing whether the semantic attribute should have the swapped values (so that users of the semantic attribute don't have to think about whether they need the swap or not). In this case, we should probably pick more descriptive names than X, Y, and Z for the semantic attribute arguments (and we'd have to add some code in to the various AST places to unswap).
> 
> WDYT?

This needs more investigation and discussion on this i believe. Both approaches will change the existing codes here and AST parts. I will prefer make this change in a separate PR since this is an issue with existing codes. We can add template support on top of old changes. @AaronBallman , would this be OK for you?

Reply:
I'm fine with that, but we should open an issue so the bug doesn't fall through the cracks.

Reply:
I have created an issue here https://github.com/intel/llvm/issues/2938 for this. Thanks.

Reply:
> Not sure about whether this will break anything. @MrSidims implemented max_work_group_size () attribute and checkWorkGroupSizeValues((),
> 
> > > > > If the attribute is only used in doing semantic checks for applying other attributes and when lowering to LLVM IR, then I think the swap should happen when lowering to LLVM IR and not when creating the semantic attribute.
> 
> could you please comment on this?

Hi, I'm on vacation somewhere in the middle of Siberia, so please don't expect fast response :) 

I agree with Aaron, that swapping order in AST is at least a confusing idea, that shall be revisited. `max_work_group_size() ` shall follow the same rules as `reqd_work_group_size()` in Sema and LLVM IR generation, hence I see no problem with moving swap.

Reply:
Also, after reviewing the code, I feel there shall be some more rules for `num_simd_work_items` attribute itteracting with `reqd_work_group_size`. I'll check it after a vacation.

Reply:
Thanks @MrSidims for the comments and checking it. Please create a separate issue about the new rules in Sema for num_simd_work_items attribute itteracting with reqd_work_group_size and let me know. I will wok on it.